front/lib/api/files/upsert.ts

```diff
@@ -0,0 +1,345 @@
+    useAppForHeaderDetection: true,
+  });
+
+  if (upsertTableRes.isErr()) {
+    return new Err({
+      name: "dust_error",
+      code: "internal_server_error",
+      message: "There was an error upserting the table.",
+      data_source_error: upsertTableRes.error,
+    });
+  }
+
+  return new Ok(undefined);
+};
+
+// Preprocessing for file upload.
+
+type PostprocessingFunction = ({
+  auth,
+  file,
+  content,
+  dataSource,
+}: {
+  auth: Authenticator;
+  file: FileResource;
+  content: string;
+  dataSource: DataSourceResource;
+}) => Promise<Result<undefined, Error>>;
+
+type PostprocessingPerUseCase = {
+  [k in FileUseCase]: PostprocessingFunction | undefined;
+};
+
+type PostprocessingPerContentType = {
+  [k in PlainTextContentType]: PostprocessingPerUseCase | undefined;
+};
+
+const processingPerContentType: PostprocessingPerContentType = {
+  "application/msword": {
+    conversation: upsertDocumentToDatasource,
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "application/vnd.openxmlformats-officedocument.wordprocessingml.document": {
+    conversation: upsertDocumentToDatasource,
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "application/pdf": {
+    conversation: upsertDocumentToDatasource,
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "text/comma-separated-values": {
+    conversation: upsertTableToDatasource,
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "text/csv": {
+    conversation: upsertTableToDatasource,
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "text/markdown": {
+    conversation: upsertDocumentToDatasource,
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "text/plain": {
+    conversation: upsertDocumentToDatasource,
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "text/tab-separated-values": {
+    conversation: upsertDocumentToDatasource, // Should it be upsertTableToDatasource?
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+  "text/tsv": {
+    conversation: upsertDocumentToDatasource, // Should it be upsertTableToDatasource?
+    avatar: notSupportedError,
+    tool_output: notSupportedError,
+  },
+};
+
+const maybeApplyProcessing: PostprocessingFunction = async ({
+  auth,
+  content,
+  file,
+  dataSource,
+}) => {
+  const contentTypeProcessing =
+    isSupportedPlainTextContentType(file.contentType) &&
+    processingPerContentType[file.contentType];
+  if (!contentTypeProcessing) {
+    return new Ok(undefined);
+  }
+
+  const processing = contentTypeProcessing[file.useCase];
+  if (processing) {
+    const res = await processing({ auth, file, content, dataSource });
+    if (res.isErr()) {
+      return res;
+    } else {
+      return new Ok(undefined);
+    }
+  }
+
+  return new Ok(undefined);
+};
+
+export async function processAndUpsertToDataSource(
```

Comment:
Can we reflect postProcessing in the name?

Reply:
I want to avoid pre/post. I actually updated everything to "processing".